### PR TITLE
clerk-auth: Migrate to Clerk Core v2

### PR DESCRIPTION
**Pull request by MonsterDeveloper**
Yesterday [Clerk Core v2](https://clerk.com/docs/upgrade-guides/core-2/overview) was released.

This PR updates the `clerk-auth` middleware to use the new version of Clerk Backend API.

I updated the middleware according to the [official Clerk's migration guide](https://clerk.com/docs/upgrade-guides/core-2/backend), as well as the tests (based on the [Clerk's Fastify middleware](https://github.com/clerk/javascript/blob/83ec173b08bdf18fda805e0d68e0034dbae0eb24/packages/fastify/src/__tests__/withClerkMiddleware.test.ts).

### CODE REVIEW

File: packages/clerk-auth/src/index.ts
Comment (octoper) [inline packages/clerk-auth/src/index.ts:2]:
```suggestion

```
It's better not to import the type from dist but infer it instead, I have left another comment bellow

**Review by octoper — CHANGES_REQUESTED** [review]
Hello @MonsterDeveloper, thank you for the amazing job with this PR!

I've left some changes that need to be made before merging it

File: packages/clerk-auth/package.json
Comment (octoper) [inline packages/clerk-auth/package.json:46]: We can change this to:
```suggestion
    "@clerk/backend": "^1.0.0",
    "@clerk/shared": "^2.0.0",
```

File: packages/clerk-auth/package.json
Comment (octoper) [inline packages/clerk-auth/package.json:53]: ❓ Do we need `rimraf` as I don't see it to be used anywhere?

**Comment by yusukebe** [comment]
@MonsterDeveloper @octoper Thanks both! I'll merge it when it is finished.

File: packages/clerk-auth/package.json
Comment (MonsterDeveloper) [reply packages/clerk-auth/package.json:41]: Maybe I'm missing something, but I thought `1` was supposed to mean any version with a restriction to a major `1` (i.e. `1.0.0` and `1.2.3`, but not `2.0.0` or `0.5.0`), which makes sense to me since we want to pin it to a current major version.

Your suggested semver `>=1.0.*` actually matches any version (including major ones like `2` and so on), starting from `1.0.0`. I suppose we can't be sure next major versions of Clerk SDK will be supported, right? Please correct me if I'm wrong.

File: packages/clerk-auth/jest.config.cjs
Comment (MonsterDeveloper) [reply packages/clerk-auth/jest.config.cjs:9]: Had to remove it due to an error:
CleanShot 2024-04-21 at 17 58 32@2x

Why do we need it though? The test pass without this mapping.

**Comment by MonsterDeveloper** [comment]
@octoper hey, thanks for your comments! Made some changes, and left some questions in the conversations. Would appreciate your review:)

File: packages/clerk-auth/jest.config.cjs
Comment (octoper) [reply packages/clerk-auth/jest.config.cjs:9]: Yeah you are right this is not needed anymore!

File: packages/clerk-auth/package.json
Comment (octoper) [reply packages/clerk-auth/package.json:41]: Yeah you are right, sorry this should be `^1.0.0`not `>=1.0.*`

**Comment by MonsterDeveloper** [comment]
@octoper seems like everything is resolved now. Are we ready to merge?

**Comment by dimkl** [comment]
Nice work @MonsterDeveloper!!

File: packages/clerk-auth/src/index.ts
Comment (octoper) [reply packages/clerk-auth/src/index.ts:51]: @dimkl  the `c.req` is `HonoRequest` which is like a more convenient wrapper.

File: packages/clerk-auth/src/index.ts
Comment (MonsterDeveloper) [reply packages/clerk-auth/src/index.ts]: Yeah that works. Done.

File: packages/clerk-auth/src/index.ts
Comment (MonsterDeveloper) [reply packages/clerk-auth/src/index.ts:50]: Yep. `c.req` is `HonoRequest`, `c.req.raw` is Web Standard `Request`.

**Review by octoper — APPROVED** [review]

**Pull request merged**
